### PR TITLE
Code coverage checks and reports

### PR DESCRIPTION
**Pull request by hsubramanianaks**
This PR is for generating code coverage reports via [c8 library](https://github.com/bcoe/c8) and uploading them via codecov github action [here](https://github.com/codecov/codecov-action)

image

image

Sample Code Cove report

https://app.codecov.io/github/hsubramanianaks/vscode-bridge-to-kubernetes/commit/32888d1f7e1a2ef1ff5f3e5ddbd93c242092e6b9/tree/src/clients

Also PR comment will be generated by codecov stating Code Coverage reports [here](https://github.com/hsubramanianaks/vscode-bridge-to-kubernetes/pull/1#issuecomment-1662707461)

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -31,6 +31,13 @@ jobs:
     - name: Smoke test (Mac, Windows)
       run: npm run test
       if: runner.os != 'Linux'
+    - name: Upload coverage reports to Codecov
+      uses: codecov/codecov-action@def81595a67c83cbb0a42ba9b624b97f4c7deb3d #v3.1.4
```
Comment (peterbom) [inline .github/workflows/build.yml:35]: Do we need a very specific version here? Would it work with `uses: codecov/codecov-action`?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -31,6 +31,13 @@ jobs:
...
+      env:
```
Comment (peterbom) [inline .github/workflows/build.yml:36]: In [here](https://github.com/codecov/codecov-action#usage) it suggests putting these options in a `with` section, rather than an `env` section (also you'd set `token` instead of `CODECOV_TOKEN`). Do you reckon we should follow that instead?

File: src/connect/ConnectWorkspaceFolder.ts [hunk 3]
```diff
@@ -398,12 +398,7 @@ export class ConnectWorkspaceFolder extends WorkspaceFolderBase {
 
         try {
             // Send POST request to http://localhost:<control-port>/api/remoting/stop.
-            await request.post({
-                method: `POST`,
-                uri: `http://localhost:${this._controlPort.toString()}/api/remoting/stop/`,
-                body: ``,
-                simple: true
-            });
+            await got.post(`http://localhost:${this._controlPort.toString()}/api/remoting/stop/`);
```
Comment (peterbom) [inline src/connect/ConnectWorkspaceFolder.ts:401]: Is this change required to make tests pass? Does the `got` library allow mock http requests?

File: .github/workflows/build.yml (hunk 1 above)
Comment (hsubramanianaks) [reply .github/workflows/build.yml:35]: Yes for security purposes we ping to sha so that even if other sha's have any security vulnerabilities or if this particular commit has security issues we will know easily and this has been followed across AKS atlanta team repo's.

File: .github/workflows/build.yml (hunk 2 above)
Comment (hsubramanianaks) [reply .github/workflows/build.yml:36]: good point, I missed it. Thanks, I will change it soon.

File: src/connect/ConnectWorkspaceFolder.ts (hunk 3 above)
Comment (hsubramanianaks) [reply src/connect/ConnectWorkspaceFolder.ts:401]: this is part of another PR , I worked on top of it. Sorry for the confusion. Here is the link https://github.com/Azure/vscode-bridge-to-kubernetes/pull/61

**Comment by Tatsinnit** [comment]
💡🙏 Thank you so much for this, Here is my 2 cents: please note some key points if we plan to make this `codecov` public that we stay true to the actual reality and future plans around it for how to.

Like:

* What is the current whole repo level coverage number?
* How do we plan to work around it, as this was bunch of already existing and matured code we could think of only covering key assets which are core of this tool and make that as a key measure. WDYT?
* Or the intent behind is to not make it visible but just to keep this at a dir level check? In that case how would that fit if change some file outside this dir, will it be triggering these checks all the time?

[Reference to the codecov for this work](https://app.codecov.io/github/hsubramanianaks/vscode-bridge-to-kubernetes/commit/32888d1f7e1a2ef1ff5f3e5ddbd93c242092e6b9/tree/src/clients)
 
Screenshot 2023-08-04 at 6 36 26 AM

File: src/tests/TestsRunHost.ts [hunk 4]
```diff
@@ -3,23 +3,26 @@
 // ----------------------------------------------------------------------------
 'use strict';
 
+import {
+    runTests
+} from '@vscode/test-electron';
 import * as path from 'path';
```
Comment (Tatsinnit) [inline src/tests/TestsRunHost.ts:9]: 💡 Just as a note: I see this kind of really solves the mystery for me, essentially all these tests already existed, and upon adding the test runner and coverage we get 63% coverage for free, which is a win! 🐴🚀🔔

**Comment by hsubramanianaks** [comment]
> 💡🙏 Thank you so much for this, Here is my 2 cents: please note some key points if we plan to make this `codecov` public that we stay true to the actual reality and future plans around it for how to. I guess this is great place where we can learn for various other project which might have similar issue aka well developed and matured and since they were legacy the codecov was not the initial thinking process. ☕️
> 
> Like:
> 
> * ~What is the current whole repo level coverage number?~
>   * **Answer** the whole repo is `+63%` which is not bad: [Details link here](https://app.codecov.io/github/hsubramanianaks/vscode-bridge-to-kubernetes/commit/32888d1f7e1a2ef1ff5f3e5ddbd93c242092e6b9/tree/src)
> 
> [Reference to the codecov for this work](https://app.codecov.io/github/hsubramanianaks/vscode-bridge-to-kubernetes/commit/32888d1f7e1a2ef1ff5f3e5ddbd93c242092e6b9/tree/src/clients)
> 
> Screenshot 2023-08-04 at 6 36 26 AM

@Tatsinnit thanks for your thoughts. My thought was to add this code cov at least not to reduce the existing 63% for future upcoming PR's and add a work item to increase the coverage for existing unit tests and for files which are missing unit test cases. I agree lot of this was well developed and matured but there is also lot of code which is legacy and some parts of it is not used. we can use this opportunity to refactor them and remove unused/dead code. Please let me know your thoughts further on this.

File: .github/workflows/build.yml (hunk 1 above)
Comment (hsubramanianaks) [reply .github/workflows/build.yml:35]: @peterbom more about pinning to commit sha [here](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions) 

```
Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps mitigate the risk of a bad actor adding a backdoor to the action's repository, as they would need to generate a SHA-1 collision for a valid Git object payload. When selecting a SHA, you should verify it is from the action's repository and not a repository fork.

```

**Review by bfoley13 — APPROVED** [review]
Nice to see you were able to use the codecov here as well!

**Review by peterbom — APPROVED** [review]
